File: implementation/scipamato/public/web/src/test/java/ch/difty/scipamato/publ/web/WicketTest.java
package ch.difty.scipamato.publ.web;

import static org.mockito.Mockito.when;

import java.util.Locale;

import de.agilecoders.wicket.extensions.markup.html.bootstrap.form.select.BootstrapMultiSelect;
import org.apache.wicket.markup.head.ResourceAggregator;
import org.apache.wicket.markup.head.filter.JavaScriptFilteredIntoFooterHeaderResponse;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.markup.html.form.TextField;
import org.apache.wicket.util.tester.WicketTester;
import org.junit.jupiter.api.BeforeEach;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.ApplicationContext;
import org.springframework.test.util.ReflectionTestUtils;

import ch.difty.scipamato.common.DateTimeService;
import ch.difty.scipamato.common.navigator.ItemNavigator;
import ch.difty.scipamato.common.web.ScipamatoWebSessionFacade;
import ch.difty.scipamato.publ.ScipamatoPublicApplication;
import ch.difty.scipamato.publ.persistence.api.KeywordService;
import ch.difty.scipamato.publ.persistence.api.PublicPaperService;

@SpringBootTest
@SuppressWarnings("SameParameterValue")
public abstract class WicketTest {

    @Autowired
    private ScipamatoPublicApplication application;

    @Autowired
    private ApplicationContext applicationContextMock;

    @SuppressWarnings("unused")
    @Autowired
    private DateTimeService dateTimeService;

    @MockBean
    private ScipamatoWebSessionFacade sessionFacadeMock;

    @MockBean
    private ItemNavigator<Long> itemNavigatorMock;

    @MockBean
    private PublicPaperService paperServiceMock;

    @MockBean
    private KeywordService keywordServiceMock;

    private WicketTester tester;

    public WicketTester getTester() {
        return tester;
    }

    protected ItemNavigator<Long> getItemNavigator() {
        return itemNavigatorMock;
    }

    protected PublicPaperService getPaperService() {
        return paperServiceMock;
    }

    @BeforeEach
    public final void setUp() {
        application.setHeaderResponseDecorator(
            r -> new ResourceAggregator(new JavaScriptFilteredIntoFooterHeaderResponse(r, "footer-container")));

        ReflectionTestUtils.setField(application, "applicationContext", applicationContextMock);
        tester = new WicketTester(application);
        when(sessionFacadeMock.getPaperIdManager()).thenReturn(itemNavigatorMock);
        Locale locale = new Locale("en_US");
        when(sessionFacadeMock.getLanguageCode()).thenReturn(locale.getLanguage());
        getTester()
            .getSession()
            .setLocale(locale);
        setUpHook();
    }

    /**
     * override if needed
     */
    protected void setUpHook() {
    }

    protected void assertLabeledTextField(String b, String id) {
        final String bb = b + ":" + id;
        getTester().assertComponent(bb + "Label", Label.class);
        getTester().assertComponent(bb, TextField.class);
    }

    protected void assertLabeledMultiSelect(String b, String id) {
        final String bb = b + ":" + id;
        getTester().assertComponent(bb + "Label", Label.class);
        getTester().assertComponent(bb, BootstrapMultiSelect.class);
    }
}

File: public/public-web/src/test/kotlin/ch/difty/scipamato/publ/web/WicketTest.kt
package ch.difty.scipamato.publ.web

import ch.difty.scipamato.common.DateTimeService
import ch.difty.scipamato.common.navigator.ItemNavigator
import ch.difty.scipamato.common.persistence.paging.PaginationContext
import ch.difty.scipamato.common.web.AbstractPage
import ch.difty.scipamato.common.web.ScipamatoWebSessionFacade
import ch.difty.scipamato.publ.ScipamatoPublicApplication
import ch.difty.scipamato.publ.entity.Keyword
import ch.difty.scipamato.publ.persistence.api.KeywordService
import ch.difty.scipamato.publ.persistence.api.PublicPaperService
import com.ninjasquad.springmockk.MockkBean
import de.agilecoders.wicket.extensions.markup.html.bootstrap.form.select.BootstrapMultiSelect
import io.mockk.Matcher
import io.mockk.MockKMatcherScope
import io.mockk.every
import org.apache.wicket.markup.head.filter.JavaScriptFilteredIntoFooterHeaderResponse
import org.apache.wicket.markup.html.basic.Label
import org.apache.wicket.markup.html.form.TextField
import org.apache.wicket.util.tester.WicketTester
import org.junit.jupiter.api.BeforeEach
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.ApplicationContext
import org.springframework.test.util.ReflectionTestUtils
import java.util.Locale

@Suppress("UnusedPrivateMember", "UnnecessaryAbstractClass")
@SpringBootTest
abstract class WicketTest {

    @Autowired
    private lateinit var application: ScipamatoPublicApplication

    @Autowired
    private lateinit var applicationContextMock: ApplicationContext

    @Autowired
    private lateinit var dateTimeService: DateTimeService

    @MockkBean
    private lateinit var sessionFacadeMock: ScipamatoWebSessionFacade

    @MockkBean(relaxed = true)
    protected lateinit var itemNavigator: ItemNavigator<Long>

    @MockkBean(relaxed = true)
    protected lateinit var paperService: PublicPaperService

    @MockkBean(relaxed = true)
    private lateinit var keywordServiceMock: KeywordService

    lateinit var tester: WicketTester
        private set

    @BeforeEach
    fun setUp() {
        application.headerResponseDecorators.add { r ->
            JavaScriptFilteredIntoFooterHeaderResponse(r, AbstractPage.FOOTER_CONTAINER)
        }
        ReflectionTestUtils.setField(application, "applicationContext", applicationContextMock)
        tester = WicketTester(application)
        every { sessionFacadeMock.paperIdManager } returns itemNavigator
        val locale = Locale("en_US")
        every { sessionFacadeMock.languageCode } returns locale.language
        tester.session.locale = locale
        every { keywordServiceMock.findKeywords("en_us") } returns listOf(
            Keyword(10, 1, "en", "k1", null),
            Keyword(11, 2, "en", "k2", null)
        )
        setUpHook()
    }

    /**
     * override if needed
     */
    protected open fun setUpHook() {}

    protected fun assertLabeledTextField(b: String, id: String) {
        val bb = "$b:$id"
        tester.assertComponent(bb + "Label", Label::class.java)
        tester.assertComponent(bb, TextField::class.java)
    }

    @Suppress("SameParameterValue")
    protected fun assertLabeledMultiSelect(b: String, id: String) {
        val bb = "$b:$id"
        tester.assertComponent(bb + "Label", Label::class.java)
        tester.assertComponent(bb, BootstrapMultiSelect::class.java)
    }
}

inline fun <reified T : PaginationContext> MockKMatcherScope.matchPaginationContext(
    offset: Int,
    pageSize: Int,
    sort: String
): T = match(PaginationContextMatcher(offset, pageSize, sort))

data class PaginationContextMatcher(
    val offset: Int,
    val pageSize: Int,
    val sort: String
) : Matcher<PaginationContext> {
    override fun match(arg: PaginationContext?): Boolean =
        arg != null && arg.offset == offset && arg.pageSize == pageSize && sort == arg.sort.toString()
}
